Device updates

Zainrax opened the pull request:
Includes current production changes by @hardiesoft

Additional changes primarily in [this commit](https://github.com/TheCacophonyProject/cacophony-web/commit/94b204d82eeded5f9fb9701b107b8e2f511af88f):
- New /reregister-authorized post to allow users to reassign a device back to a group with a new password (must have the users token, and the device token)
- Added extractJwtAuthorizedUserFromBody to get a authorized token from request body
- Change password hasing for devices from validation to beforeModify so that it is rehashed on modification
- allow /:id/settings get with authorised device token

arthurmcgregor commented:
I've added the scrollable menu to the animal list (to match the device list), but it now means that you can't see all of the options - in particular interesting and false-positive are at the bottom. Perhaps let me know the most common animals/tags and I'll put them at the top of the list for easy clickability.

mjs commented:
I can't imagine we'll want to query false-positives directly that often so that can stay down the bottom. Similarly non-mammals like birds and spiders can stay down (I don't know why we even bother having a spider tag!).

Perhaps start the list in this order: "interesting", possum, rat, stoat, ferret, weasel, hedgehog, cat
After that it doesn't matter so much aside from the low priority ones already mentioned being right at the bottom.

mjs commented:
Could this be the event issue: https://stackoverflow.com/questions/26402033/missing-click-event-for-span-inside-button-element-on-firefox#26403202

arthurmcgregor commented:
Great thanks - that fixed the Firefox bug :-) I've also done a few other minor fixes. 

I now think this is ready to go, so long as nothing else has slipped through in my code re-write. Will close #81 in my opinion 👍

mjs commented:
One issue is that the dropdown appears to be have some residual state when you make the next selection. For example:

![image](https://user-images.githubusercontent.com/14993/42349093-a7d02cc8-80ff-11e8-8f6d-a6e3f1070bc2.png)

Here I've just selected livingsprings02 and when I go to select the next device the list is still filtered based on what I had typed before even though the text box is empty. You then have to type something else and then backspace it to get back to the full list.

mjs commented:
Is it possible to add a little space between the selected items? They current end up right up against each other.

![image](https://user-images.githubusercontent.com/14993/42349230-1da59b72-8100-11e8-9336-48dd34865e64.png)

mjs commented:
I've just been using the changes in this PR for a while and they're great. This is a much easier way to find what you want.

mjs commented on views/getRecordings.pug:
```diff
@@ -34,11 +36,41 @@ html
+              .dropdown-item rabbit
+              .dropdown-item spider
+              .dropdown-item other
+              .dropdown-divider
```
nice idea to put the dividers in

hardiesoft committed:
- Updated bootstrap-vue-next lib
- Updated Vue3
- Reworking mask regions component to disallow self-intersecting regions, and allow an apron so regions can be created slightly outside the camera frame.
- Creating wizard steps for manual uploads
- Moving device status out of modal

hardiesoft committed:
- Fix mask regions tests

hardiesoft committed:
- Fix mask regions tests (2, linting)

hardiesoft committed:
- Fix sunset/sunrise in visits breakdown
- Add more info to mobile device cards in DevicesView

hardiesoft committed:
- Fix reference photo component, make it work on mobile.
- More responsive layouts for device setup.

hardiesoft committed:
- Reference photos are properly scaled when overlaying on CptvPlayer on mobile.
- Overflowing tab list component to make longer tabbed nav lists work better on mobile.

hardiesoft committed:
- Trail cameras use an appropriate daytime image as an automatic reference image.
- Masked regions are now used when adding tracks via the API to ignore tracks that are exclusively inside the masked region.
- Use common track rect drawing code for showing tracks for trailcam images.

hardiesoft committed:
- Added a dummy track to trailcam recordings where there is no track – the actual track is created the first time a user classification is added.  This could be extended to empty CPTV files.

hardiesoft committed:
- Styling dummy track

hardiesoft committed:
- Fix tests

hardiesoft committed:
- Added ability to add a dummy track to a CPTV file that has no tracks.
- Automatically add "missed track" tag to CPTV recordings that get manual IDs when they had no tracks.
- Poll every 30 seconds when an open recording is still processing, and update UI when processing is complete.

hardiesoft committed:
- Improved behaviour when mashing on next/prev buttons in CPTV player very quickly :)

hardiesoft committed:
- Fix null settings read in mask-regions API

hardiesoft committed:
- Remove stations wording on Dashboard

hardiesoft committed:
- API Documentation fix
- Add "dummy-track" to tracker version to make it available to processing.

hardiesoft committed:
- Don't send algorithm with tracks, use default

hardiesoft committed:
- Fix mask-regions api read

hardiesoft committed:
- Clone rather than always update in place DeviceHistory entries for adding reference images and mask regions in the same location.
- Fix route for Visits modal from activity view to include selected track.

hardiesoft committed:
- Clone rather than always update in place DeviceHistory entries for adding reference images and mask regions in the same location.
- Fix route for Visits modal from activity view to include selected track. (2)

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven (WIP)

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven (WIP)
- Fix mask-regions api

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven
- Hack in support for "Animal" class from Megadetector
- Show processing progress for visits in API
- Reflect processing progress in the UI.

hardiesoft committed:
- Add UI to visits when recordings are still processing
- Handle scaling DPI when drawing track boxes on trailcam images

hardiesoft committed:
- Make sure player inherits query params from parent component
- Make recording/visits exports work.

hardiesoft committed:
- Don't delete previous reference images

Zainrax committed:
add reregister-authorized, change bcrypt hasing on device to beforeModify

Zainrax committed:
lint fix

hardiesoft reviewed (approved):
I'll approve this now, but I'll probably re-work the jwt extraction middleware in the not-too-distant future so that we can more easily share that logic rather than having to copy/paste it just because the source field is different.  Looks good, thanks!  Seems like this should make device management much more logical for users!
